### PR TITLE
Added Purpose section to DS introduction

### PR DESCRIPTION
**Pull request by psatyal**
Hi team, 

I've added a "Purpose" section to the introduction (preview below). This PR is to discuss this text, decide what should and should not be documented in Storybook update this text based on our shared understanding/agreement.

Thoughts, @wielinde, @b12f, @marcalcobe, @Kharonus?

> # Design System
> 
> At OpenProject, we use a design system to ensure that our design delivers a consistent experience to users. The system describes the styles, components and patterns that come together to define the overall user interface. To do this, we provide documentation explaining how each element should be used, the different states, variations and options it offers, along with example implementation.
> 
> Our goal is to reach a point where every view in OpenProject is built with elements described in the design system. However, we are aware that reaching this goal will take time for a tool as complex and layered as OpenProject.
> 
> As such, the design system is still in its infancy and will grow with each release.
> 
> ## Purpose (what it is and what it's not)
> 
> This Storybook-based design system documentation only seeks to describe the actual reusable components; it does not seek to document features where these components are used.
> 
> The purpose of this document is to:
> 
> - Have a single point of truth on what reusable components should look like, how they should behave, and what options they provide.
> - Allow product designers and front-end developers to have a shared understanding of how these common elements are specified and implemented
> - Allow QA a common base of reference when testing implementation of features using these common elements
> - Give new or outside developers important context about available components and how to use them

### CODE REVIEW

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [inline frontend/src/stories/Introduction.stories.mdx]: @psatyal I struggle with the _absoluteness_ of words such as "ensure". The DS cannot ensure it. Only we people developing products can try to do so. The DS is a toolset to make our work easier.

Also, I see the DS not only applied to _user interface_ but also to a general _user experience_. 

```suggestion
At OpenProject, we use a design system to foster that our design delivers a consistent experience to users. The system describes the styles, components and patterns that build the foundation for a consistent user experience. To do this, we provide documentation explaining how each element should be used, the different states, variations and options it offers, along with example implementation.
```

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [inline frontend/src/stories/Introduction.stories.mdx]: @psatyal Again, I would not go so far. I see no reason to define every special case here. Only, if we think that there is a general pattern that we want to see reused, I'd add it to the DS.

```suggestion
Our goal is to reach a point where parts of views in OpenProject is built with elements described in the design system. However, we are aware that reaching this goal will take time for a tool as complex and layered as OpenProject.
```

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [inline frontend/src/stories/Introduction.stories.mdx]: Agreed.

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [inline frontend/src/stories/Introduction.stories.mdx]: @psatyal I struggle again with the absoluteness of "single point of truth". I replaced it with "central documentation". We want that everyone knows where to look for guidance and orientation. The DS can never have the single point of truth, simply because it is and will always be incomplete and insufficient for describing many singular cases. To put it differently, I don't think that I'd agree that every little deviant to what is available in the DS needs to go through a DS process. I fear that we stop innovating and just be busy with ourselves.

```suggestion
- Have a central documentation of reusable components specifying how they should look like, how they should behave, how they shall get used and what options they provide.
```

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [inline frontend/src/stories/Introduction.stories.mdx]: No, the full(!) specification might be too much detail for the DS. It is not what a product designer or developer needs. She needs to know what the intention behind a component is and how it should and should not get used. For the implementation we can look into the code. Having a full specification, one that suffices for implementation and QA testing is not a light task. It creates huge costs. Currently we don't have a central specification of OP, not even if you take all software tests together. I believe we should not lightly slip into the overhead in creating such a specification. 

```suggestion
- Allow product designers and front-end developers to have a shared understanding of how the intention behind these common elements and how they should and should not get used.
```

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [inline frontend/src/stories/Introduction.stories.mdx]: I like the idea that QA should be able to check if a new feature violates what was documented in the DS. However, if the feature introduces something new, it should not directly mean that it is wrong.

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [inline frontend/src/stories/Introduction.stories.mdx]:
```suggestion
- Give developers important context about available components and how to use them, so that they can design and develop large parts of a feature independently.
```

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [inline frontend/src/stories/Introduction.stories.mdx]: As stated above, I struggle with the concept of "single point of truth". As much as I support standardization I would not go so far and call it a "single point of truth". Our products can and should be inconsistent in specific use cases.

**Comment by wielinde** [comment]
@psatyal Thank you for the proposal. It provides a great base for a healthy discussion.

Further points I would take into the purpose section:
- Having a beautiful, appealing, delightful design that is a joy to use.
- Consistency and predictability
- Less code to maintain (Comes with re-usability)
- Improved quality (A component that gets used in many places is likely to have gone through so many use cases that its quality was continuously improved with every reuse)
- Central improvements (Example: Changing one color definition will improve the contrast ratio in multiple places)
- Accessibility (When reusing components a developer or product designer shall know what is needed to provide accessibility)
- Accepted deviants from the DS (Trade offs):
  - Speed up development and reduce maintenance costs by using external libraries that do not fully follow or are not adoptable to our DS.

Maybe we could add another section called "Roadmap" where we agree on long term goals such as:
- Support for dark and high contrast themes that can be set on individual base, so that we can all take that into account for the future when we take architecture decisions, now.
- Examples for Angular, Rails and pure HTML implementations.
- While we want the DS to be composed/bootstrapped by components of itself we still want to allow enterprise clients to overwrite key design characteristics (mostly colors and borders) to significantly but not full adopt the look of OpenProject to their own corporate identity.

**Comment by psatyal** [comment]
Linking with WP: https://community.openproject.org/projects/openproject/work_packages/47724/activity

File: frontend/src/stories/Introduction.stories.mdx
Comment (b12f) [reply frontend/src/stories/Introduction.stories.mdx]: How about "help ensure"? "foster" makes no sense in this sentence I think.

File: frontend/src/stories/Introduction.stories.mdx
Comment (b12f) [reply frontend/src/stories/Introduction.stories.mdx]: I can see why "every view" is too strong,  but "parts of views" is too weak imo.

How about

> Our goal is to reach a point where every view in OpenProject has been built with our design system principles and patterns in mind.

File: frontend/src/stories/Introduction.stories.mdx
Comment (b12f) [reply frontend/src/stories/Introduction.stories.mdx]: This is mostly a naming thing, I think. "Central documentation" basically means single point of truth, no?

I also think it should not be a documentation of what they *should* look like, but what they *do* look like, but I guess this ties in to our question on whether we want to write down design specifics instead of specifying those in tickets and letting the designs speak for themselves in stories.

File: frontend/src/stories/Introduction.stories.mdx
Comment (b12f) [reply frontend/src/stories/Introduction.stories.mdx]: > - Give developers important context about available components and how to use them. We want to enable both team- and community members to design and develop complex interfaces easily, while keeping fragmentation and debt in check.

File: frontend/src/stories/Introduction.stories.mdx
Comment (b12f) [reply frontend/src/stories/Introduction.stories.mdx]: We came up with this to make clear that neither WPs nor Figma have the final say on reusable components, styles, and patterns. Instead, by documenting something and implementing it in code you can apply a PR review process to it (e.g. both one designer and one dev need to approve changes). This should make the contents of Storybook / the DS docs the best-vetted resource, thus the primary tool to use when to decide what UI elements to use and how to use them.

Of course you'll have specific one-off instances of UI. For example, there'll always only be one app header or one global search bar, but even those should use the same colors, typo and spacings as the rest of the app.

I'm fine with renaming tho, SPOT was chosen / stayed in use because it was the one name out of all the ideas that had the least resistance among the team.

File: frontend/src/stories/Introduction.stories.mdx
Comment (oliverguenther) [reply frontend/src/stories/Introduction.stories.mdx]: I fail to see the core difference in your suggestion. Of course we will not be able to add a full specification of how a component might get used in feature X or Y. But we should strive for a complete picture on the component in isolation. This complete picture is derived from the specification and implementation rounds we have while spotifying a component. The result of it should absolutely make it into the DS

File: frontend/src/stories/Introduction.stories.mdx
Comment (oliverguenther) [reply frontend/src/stories/Introduction.stories.mdx]: > Give developers important context about available components and how to use them. We want to enable both team and community members to **compose** and develop complex interfaces easily, preventing fragmentation and technical debt where possible.

Slight variation to @b12f's input, but largely agree with their suggestion

File: frontend/src/stories/Introduction.stories.mdx
Comment (oliverguenther) [reply frontend/src/stories/Introduction.stories.mdx]: For me, the DS should be the reference I turn to when I want confirmation on behavior or design _of contained components_. This resonates with single source of truth, but of course it's not a source of truth for everything we do. It is superior over looking through a number of specification tickets that tend to be outdated after a while.

File: frontend/src/stories/Introduction.stories.mdx
Comment (marcalcobe) [inline frontend/src/stories/Introduction.stories.mdx]: Do we need to add also here "components and patterns" if we are using that in other points of the purpose?

File: frontend/src/stories/Introduction.stories.mdx
Comment (marcalcobe) [reply frontend/src/stories/Introduction.stories.mdx]: Again a point here if we use only "components" or "components and patterns" in the explanations.

About the central documentation I thing it makes it less strong using this term.

File: frontend/src/stories/Introduction.stories.mdx
Comment (marcalcobe) [reply frontend/src/stories/Introduction.stories.mdx]: But as a "common base" or "generic point" to QA to check it makes sense to have this. If a feature introduces something new probably it will be documented by the team in the work package so it is already known by the QA tester.

**Review by marcalcobe — COMMENTED** [review]
I've added some points to the discussion but I also want to stress what have been discussed with Niels in the product Daily, where we questioned the (what and what not to do) of the title if then we only talk on how to use the DS and not what are not-to-do points.

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [reply frontend/src/stories/Introduction.stories.mdx]: Agree.

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [reply frontend/src/stories/Introduction.stories.mdx]: @b12f Yes, I agree with changing it to "_do_ look like". It means that whatever is defined here in the DS is implemented and meant to be used right now. 

This also means that we do not specify here how something should look like in the future. @psatyal  I think we should put that differentiation somewhere here in the "Purpose".

@marcalcobe "Single point of truth" is much more absolute than a _documentation_ or a _catalogue_ or simply a _system_. It means that it contains all(!) the truth as it is even the single point of truth. And that is impossible to achieve. OpenProject has so much more details and decisions that are not part and will never be part of the design system. For example: If there is a feature that is a bit different to what is implemented in the DS because that deviation makes the feature better but at the same time we do not think that this will be a reusable pattern, the DS will not contain that deviation. Still, the truth is that the feature with the deviation exists. Consequently the name "single point of truth" does not work. I personally would call it a _design system_. And I would put what it means to us in this section here and not in the name.

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [reply frontend/src/stories/Introduction.stories.mdx]: I look at it from the perspective: What does a developer or product designer need to know to _reuse_ the component. That for sure should go into the DS here. However, I don't see a reason why we should repeat style definitions here in text form that are implemented in CSS if those style definitions are not needed for reusing the component/pattern. Everybody can look them up in the code or with the browser's dev tools when needed.

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [reply frontend/src/stories/Introduction.stories.mdx]: Generally I think we are all on one page here on what we want to use the DS for. 

The point that @b12f makes "...neither WPs nor Figma have the final say on reusable components, styles, and patterns" means that changing a reusable pattern should also update the DS. We should have that somewhere written down.

I believe it is just the name that does not fit 100%.

My name proposal: OpenProject Design System.

File: frontend/src/stories/Introduction.stories.mdx
Comment (psatyal) [reply frontend/src/stories/Introduction.stories.mdx]: Removed references to "single point of truth" and and replaced "should" with how things actually are. I also made this clearer in the the _this document does not_ section. 

Also use "components and patterns" instead of just "components".

File: frontend/src/stories/Introduction.stories.mdx
Comment (psatyal) [reply frontend/src/stories/Introduction.stories.mdx]: Changed to: "Allow QA a common base of reference when testing implementation of features that use these reusable components and patterns (and there is a doubt about a possible deviation)"

**Comment by psatyal** [comment]
I've updated the docs taking all previous comments into account (as far as I can tell). It's ready for round 2 of review.

Based on your last comment @wielinde, I've split the Purpose section into three sections:

1. The purpose of the design systems
2. The purpose of the design system docs
3. What the document is not

I didn't add:

- The bit about trade-offs and external libraries (I think the point isn't clear and might not be needed)
- Roadmap (I don't think this is the place for it and it would be a nightmare to maintain)

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [inline frontend/src/stories/Introduction.stories.mdx]:
```suggestion
Our goal is to reach a point where every view in OpenProject has been built with our design system principles and patterns in mind. However, we are aware that reaching this goal will take time for a tool as complex and layered as OpenProject.
```

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [inline frontend/src/stories/Introduction.stories.mdx]: @psatyal I believe something is off with the sentence in the parentheses and I don't know how to fit it :-)

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [inline frontend/src/stories/Introduction.stories.mdx]:
```suggestion
- Improve accessibilty by taking accessibilty requirements into account when designing components and patterns
```

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [inline frontend/src/stories/Introduction.stories.mdx]:
```suggestion
- Provide quality assurance personnel a common base of reference when testing features that use these reusable components and patterns (and there is a doubt about an unnecessary deviation)
```

File: frontend/src/stories/Introduction.stories.mdx
Comment (wielinde) [inline frontend/src/stories/Introduction.stories.mdx]: @psatyal 6 hash signs? Is that correct.
```suggestion
###### We recognise that UI/UX has not always been the highest priority for open-source projects. This is somewhat understandable given how open-source projects have relatively fewer design resources dedicated to it than commercial products. Our goal is to do our part to improve that situation as much as we can and document our process.
```

**Review by wielinde — APPROVED** [review]
@psatyal Cool! Thank you! I just have some minor remarks. Once they are fixed, from my point of view feel free to merge (or ask others to agree to it)

File: frontend/src/stories/Introduction.stories.mdx
Comment (psatyal) [reply frontend/src/stories/Introduction.stories.mdx]: I personally think "is" sounds better in this context tghan "has been" ("a point where something has been done" sounds less natural than "a point where something is done"), but minor issue. Committed. 

Edit: Changed it back to "is"; reads strange otherwise.

File: frontend/src/stories/Introduction.stories.mdx
Comment (psatyal) [reply frontend/src/stories/Introduction.stories.mdx]: Ohh, unsure where that came from. Simply in italics now.

File: frontend/src/stories/Introduction.stories.mdx
Comment (psatyal) [reply frontend/src/stories/Introduction.stories.mdx]: Fixed.

**Comment by psatyal** [comment]
@wielinde Thanks! I made all the necessary changes but since I accepted some from your suggestions and some in a seperate commit, I seem to have managed to introduce conflicting commits. Might need a dev's help to fix them, but I'm happy to merge this PR as-is once that's done.

**Review by wielinde — APPROVED** [review]
LGTM

**Pull request merged**
